### PR TITLE
Resolves #588: Repair records missing unsplit record suffix

### LINKED ISSUE #588: Repair record store that doesn't allow split records, but missing the unspilt record suffix
In the scenario in which you have a record store that does not allow record splitting and has no store header, it is possible that a write to that store could subsequently update the header to the latest record format version -- which indicates that every record MUST end with an unspilt record suffix.   In this situation, you will end up with a store containing a mix of records without the suffix and some with the suffix, rendering the store unreadable.   We need the ability to repair this by adding back the suffix.

### PR DESCRIPTION
**Pull request by scgray**
This adds a new method to FDBRecordStore to allow you to iteratively
repair a record store that is in the following state:

- The store does not allow split records
- The format version indicates unsplit records must have a suffix indicating as such
- There are records that are missing this suffix

The logic will identify those records missing the suffix and attempt to
tack it back on.

As a note: this repair logic can only work with primary key expressions
in which the record key is always a fixed length. As such, it adds a method
to `KeyExpression` to allow you ask the question "are you fixed length"?
Many of the changes in this PR are to add this method to all of the expressions.

### CODE REVIEW

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/metadata/expressions/KeyExpression.java [hunk 1]
```diff
@@ -119,13 +119,23 @@
...
     int getColumnSize();
 
+    /**
+     * Returns whether or not the number of columns produced by the evaluation of a key expression can
+     * vary based upon data in the record itself.  Specifically, this will return {@code false} for
+     * expressions involving {@link FanType#Concatenate}.
+     *
+     * @return whether or not the column count is fixed or may vary based upon the data in the record
+     */
+    boolean isFixedColumnSize();
```
Comment (alecgrieser) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/metadata/expressions/KeyExpression.java]: Maybe I don't understand this method, but aren't the number of columns (as the term is used in `getColumnSize`) fixed for `FanType.Concatenate` as well? In particular, I think it produces one (larger than usual) column that combines all of the elements of the repeated field.

Comment (MMcM) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/metadata/expressions/KeyExpression.java]: Right. The primary key will have a nested tuple for that column. So perhaps I was mistaken and there aren't any cases where the primary key isn't a tuple of the same length as `getColumnSize`.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java [hunk 2]
```diff
@@ -2954,6 +2956,121 @@ public void validateMetaData() {
...
+     *         a value other than {@link SplitHelper#UNSPLIT_RECORD}, the {@link FDBStoreTimer.Counts#INVALID_SPLIT_SUFFIX}
+     *         counter will be incremented.
+     *     <li>For record stores in which record splitting is disabled, but the record key is longer than expected,
+     *         the {@link FDBStoreTimer.Counts#INVALID_KEY_LENGTH} counter will be incremented.
+     * </ul>
+     *
+     * @param continuation continuation from a previous repair attempt or {@code null} to start from the beginning
+     * @param scanProperties properties to provide scan limits on the repair process
+     *   record keys should be logged
+     * @return a continuation or {@code null} if the repair has been completed
```
Comment (MMcM) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java]: "a future that completes to ..."

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java [hunk 3]
```diff
@@ -2954,6 +2956,121 @@ public void validateMetaData() {
...
+     */
+    public CompletableFuture<byte[]> repairRecordKeys(@Nullable byte[] continuation, @Nonnull ScanProperties scanProperties) {
```
Comment (MMcM) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java]: Should this have a dry-run flag?

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java [hunk 4]
```diff
@@ -2954,6 +2956,121 @@ public void validateMetaData() {
...
+    private void repairRecordKeyIfNecessary(@Nonnull FDBRecordContext context, @Nonnull Subspace recordSubspace, @Nonnull KeyValue keyValue) {
+        final RecordMetaData metaData = metaDataProvider.getRecordMetaData();
+        final Tuple recordKey = recordSubspace.unpack(keyValue.getKey());
+
+        // Ignore version key
+        if (metaData.isStoreRecordVersions() && isMaybeVersion(recordKey)) {
+            return;
+        }
+
+        final Message record = serializer.deserialize(metaData, recordKey, keyValue.getValue(), getTimer());
```
Comment (MMcM) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java]: Will any serializers be confused by being given the whole key as the primary key?

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/metadata/expressions/KeyExpression.java (hunk 1 above)
Comment (scgray) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/metadata/expressions/KeyExpression.java]: Ok. I can remove this.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java [hunk 5]
```diff
@@ -2954,6 +2956,121 @@ public void validateMetaData() {
...
+            if (!(suffix instanceof Long) || !(((Long) suffix) == SplitHelper.UNSPLIT_RECORD)) {
+                context.increment(FDBStoreTimer.Counts.INVALID_SPLIT_SUFFIX);
+                if (LOGGER.isDebugEnabled()) {
+                    LOGGER.debug(KeyValueLogMessage.of("Invalid split suffix",
+                            subspaceProvider.logKey(), subspaceProvider.toString(context),
+                            LogMessageKeys.RECORD_TYPE, recordType.getName(),
+                            LogMessageKeys.PRIMARY_KEY, recordKey));
+                }
+            }
+        } else if (recordKey.size() > primaryKeyExpression.getColumnSize() + 1) {
```
Comment (MMcM) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java]: Why not just `else`? What is the remaining legit case?

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java (hunk 3 above)
Comment (scgray) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java]: I had started to add one, but I had also tried to have it log everything that it found and it got to be a lot of code to manage not "over-logging", but now that you raise this again, I realized that I could at least still have it incrementing the counters but not doing the correction. I'll add it back.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java (hunk 2 above)
Comment (scgray) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java]: Fixed.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java (hunk 3 above)
Comment (scgray) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java]: Done.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java (hunk 4 above)
Comment (scgray) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java]: That's a good question, but I'm not sure what I can do here; I could go with the approach that, if the last element of the tuple is a `0L`, then lop it off and treat that as the PK. However, until I know the type of record it is I cannot be sure as to whether that `0L` was actually part fo the primary key, but the split suffix was missing or if it was a legit split suffix.  I'm open to suggestions on which approach might be safest.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java (hunk 5 above)
Comment (scgray) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java]: Good point. Fixed.

**Comment by scgray** [comment]
Oops. I forgot to add a test for dry-run mode.  I'll add it and update.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java (hunk 4 above)
Comment (MMcM) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java]: Let's stick with this and see if it becomes an issue.

**Pull request merged**